misc/peer-store/src/memory_store.rs

```diff
@@ -0,0 +1,421 @@
+                }
+                is_record_updated |= self.update_address(
+                    &info.peer_id,
+                    info.endpoint.get_remote_address(),
+                    AddressSource::DirectConnection,
+                    false,
+                );
+                if is_record_updated {
+                    return Some(Event::RecordUpdated(info.peer_id));
+                }
+                None
+            }
+            _ => None,
+        }
+    }
+
+    fn addresses_of_peer(&self, peer: &PeerId) -> Option<impl Iterator<Item = &Multiaddr>> {
+        self.address_book
+            .get(peer)
+            .map(|record| record.records().map(|r| r.address))
+    }
+
+    fn certified_addresses_of_peer(
+        &self,
+        peer: &PeerId,
+    ) -> Option<impl Iterator<Item = &Multiaddr>> {
+        self.address_book.get(peer).and_then(|record| {
+            Some(
+                record
+                    .records()
+                    .filter(|r| r.signed.is_some())
+                    .map(|r| r.address),
+            )
+        })
+    }
+
+    fn check_ttl(&mut self) {
+        let now = Instant::now();
+        for r in &mut self.address_book.values_mut() {
+            r.check_ttl(now, self.config.record_ttl);
+        }
+    }
+}
+
+impl Behaviour<MemoryStore> {
+    /// Get all stored address records of the peer.
+    pub fn address_record_of_peer(
+        &self,
+        peer: &PeerId,
+    ) -> Option<impl Iterator<Item = AddressRecord>> {
+        self.store()
+            .address_book
+            .get(peer)
+            .and_then(|r| Some(r.records()))
+    }
+}
+
+pub struct Config {
+    /// TTL for a record.
+    record_ttl: Duration,
+    /// The capacaity of a record store.  
+    /// The least used record will be discarded when the store is full.
+    record_capacity: NonZeroUsize,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            record_ttl: Duration::from_secs(600),
+            record_capacity: NonZeroUsize::try_from(8).expect("8 > 0"),
+        }
+    }
+}
+
+pub struct AddressRecord<'a> {
+    /// The last time we saw this address.
+    last_seen: &'a Instant,
+    /// The address of this record.
+    pub address: &'a Multiaddr,
+    /// How we observed the address.
+    pub source: AddressSource,
+    /// Whether the address expires.
+    pub should_expire: bool,
+    /// Whether the address is signed.
+    pub signed: Option<&'a libp2p_core::PeerRecord>,
+}
+impl AddressRecord<'_> {
+    /// How much time has passed since the address is last reported wrt. the given instant.  
+    pub fn last_seen_since(&self, now: Instant) -> Duration {
+        now.duration_since(*self.last_seen)
+    }
+    /// How much time has passed since the address is last reported wrt. current time.
+    pub fn last_seen(&self) -> Duration {
+        let now = Instant::now();
+        now.duration_since(*self.last_seen)
+    }

```

Comment:
Why not just make `last_seen` public like the rest?

Reply:
idk, they will have more semantic meanings this way I guess.